jwst/undersampling_correction/undersampling_correction.py

```diff
@@ -69,6 +69,10 @@ def flag_pixels(data, gdq, signal_threshold):
 
     lowest_exc_1d = np.zeros(num_pix) + n_grps
 
+    ncols = data.shape[3]
+    nrows = data.shape[2]
+
+    # Loop over all groups to locate exceedances
     for ii_int in range(n_ints):
```

Comment:
I know this isn't part of the current updates in this PR, but I'm wondering if there's a way to avoid all of these loops over integrations and groups (which are slow in python). Why not just take advantage of the natural looping/broadcasting ability of python and numpy to do it all in one fell swoop. For example, can't all of the initial pixels over the limit (in all integrations and all groups) just be flagged all at once using something simple like:

gdq[data > signal_threshold] = np.bitwise_or(gdq[data > signal_threshold], dqflags.group['UNDERSAMP'] | dqflags.group['DO_NOT_USE')

There's no need to check for and exclude pixels that already have UNDERSAMP set, because the logical OR is not going to double up any of the flags. The only additional check that I see as possibly necessary is to exclude pixels that already have DO_NOT_USE set from a previous step, so that the later steps like jump and ramp_fit can sort out which ones are "normal" DO_NOT_USE and which are DO_NOT_USE because of the undersampling correction (which requires special handling in ramp_fit).

So some playing around suggests that the whole thing could be accomplished using something like:

undersamp_pix = (data > signal_threshold) == (gdq != dqflags.group['DO_NOT_USE')
gdq[undersamp_pix] = np.bitwise_or(gdq[undersamp_pix], dqflags.group['UNDERSAMP'])

The first statement sets up an array that's True only where a pixel is above the threshold AND not already set to DO_NOT_USE, and then that array is used to set the new flag only for those pixels, over all 4 dimensions of the data/dq arrays at once. No looping anywhere.

Reply:
I will take a look at this. However, also to consider is that once a pixel in an integration has been  flagged due to being an exceedance in a particular group, a requirement is that all later groups for that pixel and integration must ALWAYS  also be flagged, even if they are not exceedances ( so if even data>signal_threshold is not true), which would not hold for your undersamp_pix expression.

Reply:
Ok,  assuming that the signal will always increase in later groups will simplify the code, as all groups can be considered independent of each other.

Reply:
If that was not the case we have bigger issues to solve :-)